src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -171,64 +214,242 @@ public static <T> T asInterfaceInstance(final Class<T> intfc, final MethodHandle
+                    infos.add(new LocalMethodInfo(mtDesc, DEFAULT_RETHROWS, fieldName));
+                } else {
+                    infos.add(new LocalMethodInfo(mtDesc, Stream.concat(DEFAULT_RETHROWS.stream(),
+                            Arrays.stream(thrown).map(MethodHandleProxies::desc)).distinct().toList(), fieldName));
                 }
-            };
+            }
 
-        final Object proxy;
-        if (System.getSecurityManager() != null) {
-            // sun.invoke.WrapperInstance is a restricted interface not accessible
-            // by any non-null class loader.
-            final ClassLoader loader = proxyLoader;
-            proxy = AccessController.doPrivileged(new PrivilegedAction<>() {
-                public Object run() {
-                    return Proxy.newProxyInstance(
-                            loader,
-                            new Class<?>[]{ intfc, WrapperInstance.class },
-                            ih);
+            Set<Class<?>> referencedTypes = stats.referencedTypes;
+            var loader = intfc.getClassLoader();
+            Module targetModule = newDynamicModule(loader, referencedTypes);
+
+            // generate a class file in the package of the dynamic module
+            String pn = targetModule.getName();
+            String n = intfc.getName();
+            int i = n.lastIndexOf('.');
+            String cn = i > 0 ? pn + "." + n.substring(i+1) : pn + "." + n;
+            ClassDesc proxyDesc = ClassDesc.of(cn);
+            byte[] template = createTemplate(loader, proxyDesc, desc(intfc), stats.singleName, infos);
+            var definer = new Lookup(intfc).makeHiddenClassDefiner(cn, template, Set.of(), DUMPER);
+            Lookup lookup;
+
+            @SuppressWarnings("removal")
+            var sm = System.getSecurityManager();
+            if (sm != null) {
+                @SuppressWarnings("removal")
+                var l = AccessController.doPrivileged((PrivilegedAction<Lookup>) () -> definer.defineClassAsLookup(true));
+                lookup = l;
+            } else {
+                lookup = definer.defineClassAsLookup(true);
+            }
+
+            MethodHandle constructor;
+            MethodHandle targetGetter;
+            try {
+                constructor = lookup.findConstructor(lookup.lookupClass(), MT_void_Lookup_MethodHandle_MethodHandle)
+                        .asType(MT_Object_Lookup_MethodHandle_MethodHandle);
+                targetGetter = lookup.findGetter(lookup.lookupClass(), ORIGINAL_TARGET_NAME, MethodHandle.class)
+                        .asType(MT_MethodHandle_Object);
+            } catch (Throwable ex) {
+                throw uncaughtException(ex);
+            }
+            return new SoftReference<>(new ProxyClassInfo(constructor, lookup, targetGetter));
+        }
+    };
+
+    private static ProxyClassInfo getProxyClassInfo(Class<?> intfc) throws IllegalArgumentException {
+        while (true) {
```

Comment:
I think this needs to recompute the class value only one time if it finds the proxy class is unloaded and the reference is cleared.   The caller and the locals are holding a strong reference to the interface and so the newly created proxy class won't be GC'ed.